inhibit/inhibit_test.go

```diff
@@ -140,55 +140,87 @@ func TestInhibitRuleHasEqual(t *testing.T) {
+		Equal:       model.LabelNames{"e"},
+	}
+	rule2 := config.InhibitRule{
+		SourceMatch: map[string]string{"s2": "1"},
+		TargetMatch: map[string]string{"t2": "1"},
 		Equal:       model.LabelNames{"e"},
 	}
 	m := types.NewMarker(prometheus.NewRegistry())
-	ih := NewInhibitor(nil, []*config.InhibitRule{&cr}, m, nopLogger)
-	ir := ih.rules[0]
+	ih := NewInhibitor(nil, []*config.InhibitRule{&rule1, &rule2}, m, nopLogger)
 	now := time.Now()
-	// Active alert that matches the source filter
-	sourceAlert := &types.Alert{
+	// Active alert that matches the source filter of rule1.
+	sourceAlert1 := &types.Alert{
+		Alert: model.Alert{
+			Labels:   model.LabelSet{"s1": "1", "t1": "2", "e": "1"},
+			StartsAt: now.Add(-time.Minute),
+			EndsAt:   now.Add(time.Hour),
+		},
+	}
+	// Active alert that matches the source filter _and_ the target filter of rule2.
+	sourceAlert2 := &types.Alert{
 		Alert: model.Alert{
-			Labels:   model.LabelSet{"s": "1", "e": "1"},
+			Labels:   model.LabelSet{"s2": "1", "t2": "1", "e": "1"},
 			StartsAt: now.Add(-time.Minute),
 			EndsAt:   now.Add(time.Hour),
 		},
 	}
 
-	ir.scache = store.NewAlerts(5 * time.Minute)
-	ir.scache.Set(sourceAlert)
+	ih.rules[0].scache = store.NewAlerts(5 * time.Minute)
+	ih.rules[0].scache.Set(sourceAlert1)
+	ih.rules[1].scache = store.NewAlerts(5 * time.Minute)
+	ih.rules[1].scache.Set(sourceAlert2)
 
 	cases := []struct {
 		target   model.LabelSet
 		expected bool
 	}{
 		{
-			// Matches target filter, inhibited
-			target:   model.LabelSet{"t": "1", "e": "1"},
+			// Matches target filter of rule1, inhibited.
+			target:   model.LabelSet{"t1": "1", "e": "1"},
+			expected: true,
+		},
+		{
+			// Matches target filter of rule2, inhibited.
+			target:   model.LabelSet{"t2": "1", "e": "1"},
+			expected: true,
+		},
+		{
+			// Matches target filter of rule1 (plus noise), inhibited.
+			target:   model.LabelSet{"t1": "1", "t3": "1", "e": "1"},
 			expected: true,
 		},
 		{
-			// Matches target filter (plus noise), inhibited
-			target:   model.LabelSet{"t": "1", "t2": "1", "e": "1"},
+			// Matches target filter of rule1 plus rule2, inhibited.
+			target:   model.LabelSet{"t1": "1", "t2": "1", "e": "1"},
 			expected: true,
 		},
 		{
-			// Doesn't match target filter, not inhibited
-			target:   model.LabelSet{"t": "0", "e": "1"},
+			// Doesn't match target filter, not inhibited.
+			target:   model.LabelSet{"t1": "0", "e": "1"},
 			expected: false,
 		},
 		{
-			// Matches both source and target filters, not inhibited
-			target:   model.LabelSet{"s": "1", "t": "1", "e": "1"},
+			// Matches both source and target filters of rule1,
+			// inhibited because sourceAlert1 matches only the
+			// source filter of rule1.
+			target:   model.LabelSet{"s1": "1", "t1": "1", "e": "1"},
+			expected: true,
+		},
+		{
+			// Matches both source and target filters of rule2,
+			// inhibited because sourceAlert2 matches also both the
+			// source and target filterof rule1.
```

Comment:
`filter of`

Reply:
Should this say `not inhibited`? `expected=false` is in the test, and it seems to be fitting your definition of when not to self-inhibit.

Reply:
Fixed.